[RN][Releases] Add Blogpost for 0.78 (#4507)
facebook/react-native-website: website/blog/2025-02-19-react-native-0.78.md

:::info
Android’s vector drawables are not a 1:1 replacement for libraries like `react-native-svg`. They are designed specifically for Android and do not work for iOS. If you want to have the same SVGs across all platforms, you'll have to continue using `react-native-svg`. Vector drawables merely offer performance benefits at the expense of customization.
:::

### ReactNativeFactory on iOS

With React Native 0.78 we improved the integration of React Native on iOS.

This version introduces a new class called `RCTReactNativeFactory` that allows you to create instances of React Native without the need of an AppDelegate. This should allow you to create a new version of React Native in a ViewController, for example. This simplifies dramatically the integration with Brownfield apps.

Imagine that you want to show a React Native view in a View Controller of your app. Starting from React Native 0.78, what you need to do, after installing all the dependencies as shown in [this guide](/docs/next/integration-with-existing-apps?language=apple#1-set-up-directory-structure), is to add this code:

```diff

+import React

> +import import React_RCTAppDelegate

Reviewer: ```suggestion
+import React_RCTAppDelegate
```

---
title: 'React Native 0.78 - React 19 and more'
authors: [vonovak, shubham, fabriziocucci, cipolleschi]
tags: [engineering]
date: 2025-02-19
---

# React Native 0.78 - React 19 and more

Today we are excited to release React Native 0.78!

This release ships React 19 in React Native and some other relevant features like native support for Android Vector drawables and better brownfield integration for iOS.

### Highlights

- [React 19](#react-19)
- [Towards smaller and faster releases](#towards-smaller-and-faster-releases)
- [Opt-in for JavaScript logs in Metro](#opt-in-for-javascript-logs-in-metro)
- [Better Android SVG support through Vector Drawable](#better-android-svg-support-through-vector-drawable)

> - [ReactNativeFactory on iOS](#reactnativefactory-on-ios)

Reviewer 2: Note: All of these need prefixing with the final path (links are broken from the blog index page).

Author: These were working fine locally and they are working fine in the netlify preview... I don't see why they should break when we release the blogpost. 😅 

This is the same setup for the [0.77 blogpost](https://github.com/facebook/react-native-website/blob/main/website/blog/2025-01-21-version-0.77.md?plain=1#L16-L18) and they are working there.

Reviewer 3: Those links are only visible if you click into the blog post because of the truncation block above them. But if you try clicking on any anchor links from the index page (reactnative.dev/blog), they're broken. They're also broken for some of the links in the 0.76 post too:
https://github.com/facebook/react-native-website/blob/2637718f90c931825ec3cca07a1b5be802c3838e/website/blog/2024-10-23-release-0.76-new-architecture.md?plain=1#L16

Author: ah, gotcha! thanks for the clarification!

We believe this new model will benefit every developer in the React Native ecosystem, as fewer breaking changes means a more stable framework that everyone can rely on.

### Opt-in for JavaScript logs in Metro

We've added an opt-in to restore JavaScript log streaming via the Metro dev server, [previously removed in 0.77](/blog/2025/01/21/version-0.77#removal-of-consolelog-streaming-in-metro) for Community CLI users. This is in response to user feedback, as well as reviewing where we are with our replacement offerings today.

To opt in, use the new `--client-logs` flag. This can also be aliased via an npm script for convenience.

```sh
npx @react-native-community/cli start --client-logs
```

Log streaming in Metro will still be going away in future and remains off by default. However, we intend to give developers a longer migration period to adapt to this change.

This update will also be made available in the incoming 0.77.1 minor release.

> ### Better Android SVG support through Vector Drawable

Reviewer 2: Another nit: The docs on `VectorDrawable` describes only generic vector graphics, and not necessarily the SVG format. Could be confusing since `react-native-svg` is not a part of React Native core.

Reviewer 3: The section title could be rephrased to something like "Added support for Android XML drawables", because this is net-new and not limited to just vector drawables. We support loading every drawable type (shapes, layer lists, etc.) so I don't want to imply that it's only vector drawables

Author: > Another nit: The docs on VectorDrawable describes only generic vector graphics, and not necessarily the SVG format. Could be confusing since react-native-svg is not a part of React Native core.

@huntie We have a warning box at the end to avoid confusion with `react-native-svg`! 😉 

---

@Abbondanzo I remember some back and forth with this title on the doc. I think we ended up with this title because it was unclear what XML drawables are. I'm happy to revert, if that's more accurate. I'll talk with @cortinico tomorrow.

Reviewer 4: Let's go for `Added support for Android XML drawables` as this captures also other non-vector drawable resources

  public func viewdidLoad() {
    super.viewDidLoad()
    // …
+ reactNativeDelegate = ReactNativeDelegate()
+ reactNativeFactory = RCTReactNativeFactory(delegate: reactNativeDelegate!)
+ view = reactNativeFactory.rootViewFactory.view(withModuleName: "<your module name>")
  }

}

+class ReactNativeDelegate: RCTDefaultReactNativeFactoryDelegate {

+  override func sourceURL(for bridge: RCTBridge) -> URL? {
+    self.bundleURL()
+  }
+
+  override func bundleURL() -> URL? {
+    #if DEBUG
+    RCTBundleURLProvider.sharedSettings().jsBundleURL(forBundleRoot: "index")
+    #else
+    Bundle.main.url(forResource: "main", withExtension: "jsbundle")
+    #endif
+  }
+}

```

React Native will be loaded in the View Controller as soon as you navigate to it.

This code creates an `RCTReactNativeFactory`, assigns a delegate to it, and asks it to create a `rootView` for a React Native’s view.

The delegate is defined below, and it overrides the `sourceURL` and the `bundleURL` properties to tell React Native where it can find the JS bundle to load in the view.

## Other Breaking Changes

### General

- React Native DevTools
  - Synced React Native DevTools to chromium/6613

>   - Removed FuseboxClient CDP domain

Reviewer 2: To be honest, none of these changes are user facing/impacting.

- The `chromium/6613` bump is an FYI (i.e. "you might get new features from our regular Chrome frontend bump") — it doesn't represent an API change nor a functionality change for end users.
- The `FuseboxClient` CDP domain removal will matter only to select integrators — maybe keep, but can totally be relegated to "check the full changelog".

Author: I added the first because you mention it in the original doc! 🤣 Happy to remove it.

This is actually the full changelog for what concerns the breaking changes. Also, integrators are recipients of this blogpost too, so I'd keep it here.

---
title: 'React Native 0.78 - React 19 and more'
authors: [vonovak, shubham, fabriziocucci, cipolleschi]
tags: [engineering]
date: 2025-02-19
---

# React Native 0.78 - React 19 and more

Today we are excited to release React Native 0.78!

This release ships React 19 in React Native and some other relevant features like native support for Android Vector drawables and better brownfield integration for iOS.

### Highlights

- [React 19](#react-19)
- [Towards smaller and faster releases](#towards-smaller-and-faster-releases)
- [Opt-in for JavaScript logs in Metro](#opt-in-for-javascript-logs-in-metro)
- [Better Android SVG support through Vector Drawable](#better-android-svg-support-through-vector-drawable)
- [ReactNativeFactory on iOS](#reactnativefactory-on-ios)

<!--truncate-->

## Highlights

### React 19

React 19 is now available on React Native!

> React 19 requires updating your app as we introduced some changes from React 18. For example, we removed some APIs such as `propTypes`, and you need to adjust your app to make it compatible with the new version of React.

Reviewer 5: ```suggestion
React 19 requires updating your app, as we introduced some changes from React 18. For example, we removed some APIs such as `propTypes`, and you need to adjust your app to make it compatible with the new version of React.
```

With this release, we simplified the process to enable the React Compiler in your React Native apps. [In previous versions](https://react.dev/learn/react-compiler#using-react-compiler-with-react-17-or-18), you had to install two packages: the compiler and its runtime. After those packages were installed, you had to configure a Babel plugin to enable React Compiler through Metro.

Now, you only need to install the compiler itself and to configure the Babel plugin. To learn how to enable it, you can follow our step-by-step [guide](https://react.dev/learn/react-compiler#usage-with-babel).

To verify that the compiler is running, you can open the React Native DevTools: you should see that the components that are memoized have the `Memo ✨` tag attached to them in the Component Inspector.

If you want to learn more about React Compiler, these are useful resources:

- [React Compiler](https://react.dev/learn/react-compiler) docs
- [React Compiler Deep Dive](https://www.youtube.com/watch?v=uA_PVyZP7AI)

### Towards smaller and faster releases

> WWe’re updating our release process to ship stable React Native releases more frequently in 2025.

Reviewer 6: ```suggestion
We’re updating our release process to ship stable React Native releases more frequently in 2025.
```

It will be easier for you to update the React Native version because we’ll be reducing the number of breaking changes we ship. Faster releases also means that all the bugfixes we ship internally are reaching you earlier, and you can benefit from the latest features we develop inside React Native.

> We believe this new model will benefit every developer in the React Native ecosystem, as less breaking changes means a more stable framework that everyone can rely on.

Reviewer 5: nit - grammar
```suggestion
We believe this new model will benefit every developer in the React Native ecosystem, as fewer breaking changes means a more stable framework that everyone can rely on.
```

### Opt-in for JavaScript logs in Metro

> We've added an opt-in to restore JavaScript log streaming via the Metro dev server, [previously removed in 0.77](https://reactnative.dev/blog/2025/01/21/version-0.77#removal-of-consolelog-streaming-in-metro) for Community CLI users. This is in response to user feedback, as well as reviewing where we are with our replacement offerings today.
> 

Reviewer 7: ```suggestion
We've added an opt-in to restore JavaScript log streaming via the Metro dev server, [previously removed in 0.77](/blog/2025/01/21/version-0.77#removal-of-consolelog-streaming-in-metro) for Community CLI users. This is in response to user feedback, as well as reviewing where we are with our replacement offerings today.

```

Internal link nit.

To opt in, use the new `--client-logs` flag. This can also be aliased via an npm script for convenience.

```sh
npx @react-native-community/cli start --client-logs
```

Log streaming in Metro will still be going away in future and remains off by default. However, we intend to give developers a longer migration period to adapt to this change.

This update will also be made available in the incoming 0.77.1 minor release.

### Better Android SVG support through Vector Drawable

> In React Native 0.78, we’re shipping a new way to load icons, illustrations, and other graphic elements on Android as [XML resources](https://developer.android.com/guide/topics/resources/drawable-resource). This means you can use [vector drawables](https://developer.android.com/develop/ui/views/graphics/vector-drawable-resources) for displaying vector images at any scale without losing quality, or [shape drawables](https://developer.android.com/guide/topics/resources/drawable-resource#Shape) for drawing more basic embellishments. This is all supported by the same `Image` component that you know and love. To use this feature today, you can import XML resources like any other [static resource](https://reactnative.dev/docs/next/images#static-image-resources) by referencing them in the `source` prop. Furthermore, using XML resources rather than bitmaps will also help you reduce your application size and will result in better rendering across screens with different DPI.

Reviewer 7: ```suggestion
In React Native 0.78, we’re shipping a new way to load icons, illustrations, and other graphic elements on Android as [XML resources](https://developer.android.com/guide/topics/resources/drawable-resource). This means you can use [vector drawables](https://developer.android.com/develop/ui/views/graphics/vector-drawable-resources) for displaying vector images at any scale without losing quality, or [shape drawables](https://developer.android.com/guide/topics/resources/drawable-resource#Shape) for drawing more basic embellishments. This is all supported by the same `Image` component that you know and love. To use this feature today, you can import XML resources like any other [static resource](/docs/next/images#static-image-resources) by referencing them in the `source` prop. Furthermore, using XML resources rather than bitmaps will also help you reduce your application size and will result in better rendering across screens with different DPI.
```

Internal link nit.

```js
// via require
<Image
  source={require('./img/my_icon.xml')}
  style={{width: 40, height: 40}}
/>;

// or via import
import MyIcon from './img/my_icon.xml';
<Image source={MyIcon} style={{width: 40, height: 40}} />;
```

#### Performance & Quality

> [Like all other image types](https://reactnative.dev/docs/next/images#off-thread-decoding), Android’s XML resources are loaded and inflated off the main thread so you don’t drop any frames. This means the resource is not guaranteed to display instantly but also does not prevent user input while the resource is loading. Off-thread decoding is especially important when you need to render many icons at the same time. Internal apps realized some significant performance improvements when using Android’s vector drawables.

Reviewer 7: ```suggestion
[Like all other image types](/docs/next/images#off-thread-decoding), Android’s XML resources are loaded and inflated off the main thread so you don’t drop any frames. This means the resource is not guaranteed to display instantly but also does not prevent user input while the resource is loading. Off-thread decoding is especially important when you need to render many icons at the same time. Internal apps realized some significant performance improvements when using Android’s vector drawables.
```

Internal link nit.

- They cannot be loaded over the network by Metro. If you change the directory or name of an XML resource, you will need to rebuild your Android application each time.
- They have no dimensions. By default, they will display with a 0x0 size and you need to provide a width and height for them to show up.
- They are not fully customizable at runtime; you can only control dimensions or the overall tint color but you can’t customize individual element attributes _inside_ the resource like stroke widths, border radius, or colors. These types of customizations require different variants of your XML resource.

:::info
Android’s vector drawables are not a 1:1 replacement for libraries like `react-native-svg`. They are designed specifically for Android and do not work for iOS. If you want to have the same SVGs across all platforms, you'll have to continue using `react-native-svg`. Vector drawables merely offer performance benefits at the expense of customization.
:::

### ReactNativeFactory on iOS

With React Native 0.78 we improved the integration of React Native on iOS.

> This version introduces a new class called `RCTReactNativeFactory` that allows you to create instances of ReactNative without the need of an AppDelegate. This should allow you to create a new version of ReactNative in a ViewController, for example. This simplifies dramatically the integration with Brownfield apps.

Reviewer 7: ```suggestion
This version introduces a new class called `RCTReactNativeFactory` that allows you to create instances of React Native without the need of an `AppDelegate`. This should allow you to create a new version of React Native in a `ViewController`, for example. This simplifies dramatically the integration with Brownfield apps.
```

Formatting nits.

> Imagine that you want to show a React Native view in a View Controller of your app. Starting from React Native 0.78, what you need to do, after installing all the dependencies as shown in [this guide](https://reactnative.dev/docs/next/integration-with-existing-apps?language=apple#1-set-up-directory-structure), is to add this code:

Reviewer 7: ```suggestion
Imagine that you want to show a React Native view in a View Controller of your app. Starting from React Native 0.78, what you need to do, after installing all the dependencies as shown in [this guide]/docs/next/integration-with-existing-apps?language=apple#1-set-up-directory-structure), is to add this code:
```
Internal link nit.

```diff

+import React
+import import React_RCTAppDelegate

public class ViewController {

+  var reactNativeFactory: RCTReactNativeFactory?
+  var reactNativeDelegate: ReactNativeDelegate?

  public func viewdidLoad() {
    super.viewDidLoad()
    // …
+ reactNativeDelegate = ReactNativeDelegate()
+ reactNativeFactory = RCTReactNativeFactory(delegate: reactNativeDelegate!)
+ view = reactNativeFactory.rootViewFactory.view(withModuleName: "<your module name>")
  }

}

+class ReactNativeDelegate: RCTDefaultReactNativeFactoryDelegate {

+  override func sourceURL(for bridge: RCTBridge) -> URL? {
+    self.bundleURL()
+  }
+
+  override func bundleURL() -> URL? {
+    #if DEBUG
+    RCTBundleURLProvider.sharedSettings().jsBundleURL(forBundleRoot: "index")
+    #else
+    Bundle.main.url(forResource: "main", withExtension: "jsbundle")
+    #endif
+  }
+}

```

React Native will be loaded in the View Controller as soon as you navigate to it.

> This code creates a `RCTReactNativeFactory`, it assigns a delegate to it and asks it to create a `rootView` for a React Native’s view.

Reviewer 5: ```suggestion
 This code creates an `RCTReactNativeFactory`, assigns a delegate to it, and asks it to create a `rootView` for a React Native view.

```

> The delegate is defined below and it overrides the `sourceURL` and the `bundleURL` properties to tell React Native where it can find the JS bundle to load in the view.
> 

Reviewer 7: ```suggestion
The delegate is defined below, and it overrides the `sourceURL` and the `bundleURL` properties to tell React Native where it can find the JS bundle to load in the view.

```

Nit.

## Other Breaking Changes

### General

- React Native DevTools
  - Synced React Native DevTools to chromium/6613
  - Removed FuseboxClient CDP domain
- Codegen
  - Separate component array types and command array types

### Android

> - Nullability changes: migrating `RootView` to Kotlin resulted in changes of parameter types from nullable to non nullable.
> - The following classes have been moved from public to internal, or removed, and can’t be accessed anymore:
>   - `com.facebook.react.views.text.frescosupport.FrescoBasedReactTextInlineImageViewManager`
>   - `com.facebook.react.uimanager.ComponentNameResolver`
>   - `com.facebook.react.bridge.GuardedResultAsyncTask`
>   - `com.facebook.react.uimanager.FabricViewStateManager`

Reviewer 7: ```suggestion
- Nullability changes: migrating `RootView` to Kotlin resulted in changes of parameter types from nullable to non-nullable.
- The following classes have been moved from public to internal, or removed, and can’t be accessed anymore:
  - `com.facebook.react.bridge.GuardedResultAsyncTask`
  - `com.facebook.react.uimanager.ComponentNameResolver`
  - `com.facebook.react.uimanager.FabricViewStateManager`
  - `com.facebook.react.views.text.frescosupport.FrescoBasedReactTextInlineImageViewManager`
```

Alphabetical order, wording nit.

### iOS

> - Change Image load event size info from logical size to pixel
>   - This affects only the Old Architecture

Reviewer 7: ```suggestion
- Change Image load event size info from logical size to pixel (this affects only the Old Architecture)
```

## Acknowledgements

React Native 0.78 contains over 509 commits from 87 contributors. Thanks for all your hard work!

Thanks to all the additional authors that worked on documenting features in this release post:

- [Dream11](https://github.com/dream-sports-labs) team for the thorough testing of React 19 features in React Native
- [Nicola Corti](https://github.com/cortinico) for the work on Faster Releases
- [Alex Hunt](https://github.com/huntie) for the work on the Metro logs opt-in
- [Peter Abbondanzo](https://github.com/Abbondanzo) for the work on Android XML Drawable Support

> - [Oskar Wasniewski](https://github.com/okwasniewski) for the work on the ReactNativeFactory

Reviewer 8: Thanks for mentioning me 🙌

> - [Oskar Wasniewski](https://github.com/okwasniewski) for the work on the ReactNativeFactory

Author: sorry for mispelling the name... 🤦

> - [Oskar Wasniewski](https://github.com/okwasniewski) for the work on the ReactNativeFactory

Reviewer 8: No worries, I know it's a bit difficult 😅

## Upgrade to 0.78

Please use the [React Native Upgrade Helper](https://react-native-community.github.io/upgrade-helper/) to view code changes between React Native versions for existing projects, in addition to the Upgrading docs.

To create a new project:

```
npx @react-native-community/cli@latest init MyProject --version latest
```

> If you use Expo, React Native 0.78 will be supported in Expo SDK 53 (instructions on how to update React Native inside your Expo project to 0.78.0 will be available in a separate Expo blog-post in the near future).

Reviewer 7: ```suggestion
If you use Expo, React Native 0.78 will be supported in Expo SDK 53. Instructions on how to update React Native inside your Expo project to 0.78 will be available in a separate Expo blog post in the near future.
```

Sentence nits.
